oguzkocer/gutenberg
# Call for testing: Composite build for react-native-editor

This PR updates `react-native-editor` (demo) app to use composite build and make `react-native-bridge` an included build. There are many advantages of doing that such as consistency with WPAndroid and easier integration to other projects, but I'll save the full discussion for the real PR. **The goal for this PR is to test the demo app with these changes. It's not to do a full code review.** If you want to do a full review, I'd welcome any feedback. However, please keep in mind that these changes are what arrived at while beating the build files into shape, so they are far from the end result. Even the commit messages are 🤮, but that's why the branch name starts with `try/` so 🤷‍♂️ 😄 

**High level changes:**
* I can't get composite build to work as I'd like it to when the source files are in the root, so there is a new `react-native-bridge` module in the `react-native-bridge` project. (need a better name before actual PR) There might be a way to do it, but I think having such a module is a good thing. That means the `build.gradle` file for this project got split up to `build.gradle` (project build file) and `react-native-bridge/build.gradle` (module build file)
* `react-native-bridge` is the one that integrates all the subprojects and exposes it with the `api` configuration. That means projects integrating `react-native-bridge` shouldn't need to concern about where all the minor subprojects live nor will they need to manually add dependencies to them. This approach may not make much sense if these were all remote dependencies, but I think it makes a lot of sense (unless I am missing something) with the current setup.
* `react-native-aztec/android/build.gradle` changes may be unnecessary for this result. While trying different approaches, I had some build issues and this is how I fixed them. These changes are good to have, so I am leaving them in.

**To test:**
The main focus should be on the `react-native-editor` app. What I'd like to verify is:
1. You can get it to build & run. I used [these instructions](https://github.com/oguzkocer/gutenberg/blob/master/packages/react-native-editor/README.md) and it worked for me. Here are the steps I used (all commands ran from the root `gutenberg` directory):
* `nvm use`
* `npm install`
* `npm run native start`
* While ^ is running, from another terminal window: `npm run native android`

This started the emulator and the blocks seemed to be working.
2. Make changes to the `react-native-bridge` project and verify that it's reflected in the demo app.
3. Make changes to the `react-native-editor` project and verify that it works.
4. Make changes to the JS files and verify that it works.
5. Verify any other regular flow Gutenberg developers use works as expected.
6. Anything about Hermes. I don't even know if there is something to test there, but it'd really help if you can give that a shot if there is.

I don't have enough knowledge about Gutenberg to run these (2-5) myself and that's mainly why I need help before I can work on the real PR. If you'd like to document the tests you ran, it could prove very helpful for the actual PR and for future implementation. (such as using it in WPAndroid)

**Short term plans:**
1. Fix up the commit messages, code changes etc and open a PR in `wordpress/gutenberg` repo.
2. Follow up to the hermes binary PR
3. Refactor `react-native-bridge` build files to simplify the extra flags `buildGutenbergFromSource`, `buildGutenbergMobileJSBundle` and remove any extra logic. Figure out how we want to handle `buildGutenbergFromSource` for the composite builds.

**Possible plans:**
Add a new project that does a minimum binary + composite build setup for the `react-native-bridge` project. This can serve as a testing ground for build changes as well as a very simple sample for outside contributors so they can use it to setup their own project. (not sure if there is enough value for this yet, but adding it here for documentation purposes)
